src/pybind/mgr/cephadm/module.py

```diff
@@ -2686,11 +2702,23 @@ def _apply_service_spec(self, spec: ServiceSpec) -> str:
             raise OrchestratorError((f'The maximum count_per_host allowed is {max_count}.'
                                      + ' This limit can be adjusted by changing the mgr/cephadm/max_count_per_host config option'))
 
+        ports_in_use = {}
+        for h in self._get_candidate_hosts(spec.placement):
```

Comment:
This won't work in a situation where the placement is just `count: 2` or something like that as `get_candidate_hosts` will just return back an empty list and we could still end up trying to place daemons on a host with a port conflict. I think it's probably best to just give the scheduler a full dict of all the ports in use on each host then allow it to do the filtering based on that. Sort of similar to how we filter out mon hosts if the networks don't match up. Given, unlike the mon situation, we actually want to raise some sort of warning to the user if there's a port conflict, we probably need to attempt to just do something a bit more complex.  In cases where the placement is ambiguous (e.g. just count or count combined with something else so we're still only using a subset of the hosts) we should attempt to filter out hosts and then only raise some sort of warning if the count cannot be satisfied due to a port conflict. In cases where the placement is precise (e.g. explicit hosts or by label) we should raise some sort of warning if there are port conflicts on any of the hosts that placement matches. Basically, it comes down to raising a health warning if the placement can't be satisfied due to a port conflict. In all cases, we ideally should still place the daemons on the hosts where there are no conflicts and just raise a warning that there are some hosts we wanted to place on but can't due to the conflict. Admittedly, this becomes a bit complicated as it's hard to have the scheduler tell us there's hosts we can't satisfy the placement due to conflicts on X set of hosts while also still returning the list of hosts we do actually want to place on. Then there's the additional issue that in future passes we must make sure any "port conflict" we see isn't actually just the daemon from the given service being on the host (which is obviously not a real conflict because it's the daemon we want there itself using the port). The more I think about it, the trickier this seems to get right.